### ISSUE TITLE
Downloading fails as Ubuntu 22.04 is the host machine

### ISSUE THREAD

**Issue by JustOneMoreBlock** [moby/moby#44851]
### Description

We are downloading a large file using curl in a `docker container` and always fails, and presented with the following error:

`curl: (92) HTTP/2 stream 1 was not closed cleanly: INTERNAL_ERROR (err 2)`

### Reproduce

1. Run `Ubuntu 22.04` as the `host machine`.
2. Run a `docker container` in the background — or run a `stress --vm 512 --vm-bytes 1024M`.
3. `docker run --rm --name=testing --network host -m 1024M ubuntu-scripts curl https://downloadurl.com/dl.zip -JLO# 2>&1`

The downloaded `zip file` will fail to download and be presented with this error: `curl: (92) HTTP/2 stream 1 was not closed cleanly: INTERNAL_ERROR (err 2)`

### Expected behavior

The download should be completed 100% of the time.

### docker version

```bash
root@node:/home# docker version
Client: Docker Engine - Community
 Version:           20.10.22
 API version:       1.41
 Go version:        go1.18.9
 Git commit:        3a2c30b
 Built:             Thu Dec 15 22:28:04 2022
 OS/Arch:           linux/amd64
 Context:           default
 Experimental:      true

Server: Docker Engine - Community
 Engine:
  Version:          20.10.22
  API version:      1.41 (minimum version 1.12)
  Go version:       go1.18.9
  Git commit:       42c8b31
  Built:            Thu Dec 15 22:25:49 2022
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.6.15
  GitCommit:        5b842e528e99d4d4c1686467debf2bd4b88ecd86
 runc:
  Version:          1.1.4
  GitCommit:        v1.1.4-0-g5fd4c4d
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
```
- The `docker version` doesn't seem to matter as we have tried old versions, and the error is persistent.
```


### docker info

```bash
root@node:~# docker info
Client:
 Context:    default
 Debug Mode: false
 Plugins:
  app: Docker App (Docker Inc., v0.9.1-beta3)
  buildx: Docker Buildx (Docker Inc., v0.9.1-docker)
  scan: Docker Scan (Docker Inc., v0.23.0)

Server:
 Containers: 39
  Running: 39
  Paused: 0
  Stopped: 0
 Images: 6
 Server Version: 20.10.22
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Native Overlay Diff: true
  userxattr: false
 Logging Driver: json-file
 Cgroup Driver: systemd
 Cgroup Version: 2
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: io.containerd.runtime.v1.linux runc io.containerd.runc.v2
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 5b842e528e99d4d4c1686467debf2bd4b88ecd86
 runc version: v1.1.4-0-g5fd4c4d
 init version: de40ad0
 Security Options:
  apparmor
  seccomp
   Profile: default
  cgroupns
 Kernel Version: 5.15.0-57-generic
 Operating System: Ubuntu 22.04.1 LTS
 OSType: linux
 Architecture: x86_64
 CPUs: 16
 Total Memory: 125.7GiB
 Name: hostname
 ID: XX
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 Registry: https://index.docker.io/v1/
 Labels:
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false
```

#### Dockerfile `ubuntu-base`:

```
FROM ubuntu:bionic

ENV DEBIAN_FRONTEND noninteractive

RUN apt-get update -y
RUN apt-get upgrade -y

# Install most common packages
RUN apt-get install -y \
    curl \
    ca-certificates \
    gzip \
    jq \
    python \
    unzip \
    wget

# Copy most common files
COPY init_delay /usr/bin/init_delay
RUN chmod 555 /usr/bin/init_delay
```
- The `OS` version here doesn't seem to matter as we have tried with Ubuntu 22.04, Alpine, and AlmaLinux. The error still persists.
- With different versions of the `OS`. This gave us newer versions of `curl`, and the issue persisted.

#### Dockerfile `ubuntu-scripts`:

```
FROM ubuntu-base:latest

RUN apt-get install -y \
  nano \
  locales-all \
  locales \
  bzip2 \
  nodejs \
  dos2unix \
  git \
  expect \
  file \
  iproute2 \
  p7zip \
  python3 \
  telnet \
  rsync \
  zip \
  awscli \
  openssl \
  libssl-dev \
  lftp

# Locale
ENV LC_ALL en_US.UTF-8
ENV LANG en_US.UTF-8
ENV LANGUAGE en_US.UTF-8

# Java 17 setup
RUN wget https://download.oracle.com/java/17/latest/jdk-17_linux-x64_bin.deb -O jdk-17.deb
RUN apt-get install -y ./jdk-17.deb
RUN update-alternatives --install "/usr/bin/java" "java" "/usr/lib/jvm/jdk-17/bin/java" 1

RUN ln -sf /bin/bash /bin/sh
COPY entrypoint.sh /entrypoint.sh

ENTRYPOINT [ "/bin/bash", "/entrypoint.sh" ]
```

## What are doing?

We are creating a `docker container` and simply downloading a file via `curl`. We also attempted to switch to `wget` and a similar error is produced. Keep in mind, other `docker containers` are running as well.

#### Command

```
docker run --rm --name=testing --network host -m 1024M ubuntu-scripts curl https://downloadurl.com/dl.zip -JLO# 2>&1
```
- This command can be used to re-create the issue.
- We have tried different things with `curl` to ensure there was no buffer or possible TLS issue, and this all to be proven not an issue as we have also tried many different URLs/domains/IPS, and the results are the same.
- We have tried different `docker` commands:
  - `--dns`, `--network bridge`, `--cgroupns host`

#### What's the issue?

We're running multiple `docker containers` on servers, and then we have a separate `docker image` called `ubuntu-scripts` that is created when we open up the `downloader` menu to download a file from within our control panel. This calls for the `curl` command to be executed in that `docker image` and stored in the clients' volume. Unfortunately, if the download is bigger than `1024M` then the download will `fail`.

- Running all `docker containers` and changing the `-m 1024M` to the size of the `zip file` will allow the download to succeed. This isn't ideal because if we have a `127 GB` zip file. It couldn't download that. Whereas it will succeed in that same file running on `1024M` if we stopped all containers.
- If we stop all running `docker containers` on the server, and start our testing command again. The download will succeed.
- If we start one `docker container` running in the background, and start our testing command again. The download will fail. We know we're not running out of memory.

#### Other Notable Issues

- We have servers running the same OS and Docker versions, and they're running `zfs`, and this works perfectly fine. This error seems notable on `ext4`. But, sometimes succeeds. We don't necessarily think it's an `ext4` bug.
- We can download the same zip file on the `host machine` without any issues.
- In the `docker image` testing. It's noticeable as the downloads slow down.
  - I accidently pasted the curl command in the `host machine` as I was downloading from the `testing image` while `docker containers` were running, and the download in the `testing image` sped up, and the download succeeded because I was downloading through the `host machine`. I found that interesting and strange.

#### Solutions

We sort of found a solution. But, this goes in line with the bug report.

I've previously messed around with `cgroups` prior to this. However, at the time it seemed to continue to fail. It was suggested to modify `/etc/default/grub` and added `GRUB_CMDLINE_LINUX_DEFAULT=" systemd.unified_cgroup_hierarchy=1"` and `update-grub` and to reboot the `node`.

After the reboot, I confirmed the `docker` container for the test server was online and began downloading like before, and behold the download succeeded. The difference now is we're using `cgroupsv1` whereas by default it was using `cgroupsv2`. In the `docker inspect` command it showed the following:

#### cgroupsv2
```
"CgroupnsMode": "private",
```

#### cgroupsv1
```
"CgroupnsMode": "host",
```

To ensure it was consistent, I reverted back to the default config `etc/default/grub` by removing the `systemd.unified_cgroup_hierarchy` and ran `update-grub` again, and proceeded to reboot the node. The `docker` container for the `test server` was confirmed online and running my test downloads. It continued to `"`magically`"` work 100% running `cgroupsv2`, and again I confirmed this by checking `docker inspect` was set back to `private`. I did notice minor slowdowns but always succeeded.

Now I wondered if there was a way to get this working without `rebooting` the node, and since `--cgroupns` is a flag and I knew the `cgroupsv1` is set for `host`. I attempted to run this on a different node where it was known to fail. Unfortunately, using this didn't work with the `test images` I have been running.

During another test with no servers on this node and downloading a backup was successful. I created a `test server` on this node and also downloaded it successfully without changing anything, and it's using `cgroupsv2`. However, since this node required further testing and executed `stress --vm 512 --vm-bytes 1024M` on the node and downloading failed which wasn't `docker` related. At this time, the `test server` and `stress testing` was running. As soon as I killed the `stress test` I was able to download the file successfully.

I changed the `systemd.unified_cgroup_hierarchy` to `0` in `/etc/default/grub` and ran `update-grub`, and proceeded to reboot the node. The `test server` is online. The download is successful. Re-adding in the `stress testing` again and the download is now successful.

Seems like the best fix at the moment is to change `/etc/default/grub` and add `GRUB_CMDLINE_LINUX_DEFAULT=" systemd.unified_cgroup_hierarchy=1"` and `update-grub` and reboot the node.

The bottom line seems like `docker` doesn't like `cgroupsv2`.

`╯°□°）╯︵ ┻━┻`

**Comment by AkihiroSuda** [comment]
> docker run --rm --name=testing --network host -m 1024M ubuntu-scripts curl https://downloadurl.com/dl.zip -JLO# 2>&1

- Is this reproducible with any public URL?
- Is this reproducible with the plain `ubuntu:XX.XX` image? (with `apt-get install curl`)

**Comment by JustOneMoreBlock** [comment]
```
docker pull ubuntu:bionic
docker run -d -t --name=testing --network host -m 1024M ubuntu:bionic
docker exec -it testing /bin/bash
curl http://speedtest-nyc1.digitalocean.com/5gb.test
```
- This will cause `curl` to be `command not found`.

```
root@node:/# apt-get -y install curl
Reading package lists... Done
Building dependency tree       
Reading state information... Done
E: Unable to locate package curl
```
- Unable to install

```
apt-get -y update
apt-get -y install curl
curl http://speedtest-nyc1.digitalocean.com/5gb.test
```
- Public Image
- Public URL
- Continues to fail.

---

Trying to do it in all one command `apt-get` won't install curl and ignores the docker container entirely, and downloads the file on the `host machine`.

**Comment by thaJeztah** [comment]
> Trying to do it in all one command apt-get won't install curl and ignores the docker container entirely, and downloads the file on the host machine.

Quick tip; if you need to run multiple commands, you can chain them to be the argument to `sh` (or `bash`), otherwise your shell will interpret it as `docker run <some command>` and if that is successful, run the next command (e.g. `apt-get install` or `curl).

For example; 

```bash
docker run -it --rm --name=testing --network host -m 1024M ubuntu:bionic sh -c 'apt-get update && apt-get install -y curl && curl http://speedtest-nyc1.digitalocean.com/5gb.test'
```

That run a container that runs `sh`, and run s the `apt-get update ....` inside a shell (the `-c` and quotes are important here)

**Comment by sam-thibault** [comment]
Did chaining the commands solve this issue for you?

**Comment by wezell** [comment]
+1 on this issue as this is failing for me as well using Docker Desktop 4.15.0 for OSX (Intel).  Trying to download a 1.5gb file with `curl` or `wget` fails in Ubuntu 22.04.  I tried adding `curl --http1.1 --keepalive-time 2` to the curl call with no effect.  The same curl command works on my host machine.

Additional information, the file is not being written to the docker fs, it is being downloaded to a volume I am mounting in.

**Comment by sam-thibault** [comment]
@wezell can you provide more details for your steps to reproduce this issue?  Are you mounting the volume with a bind mount?

It may be related to how Docker Desktop connects the host file system with the VM [for example with virtiofs].  In which case it may be better to open an issue for this in the https://github.com/docker/for-mac issue tracker.

**Comment by sam-thibault** [comment]
We weren't able to reproduce this, so I will close it as stale.  We can reopen if there are comments with more information.
